consensus/istanbul/backend/announce.go

```diff
@@ -1199,12 +967,12 @@ func (sb *Backend) setAndShareUpdatedAnnounceVersion(version uint) error {
 	}
 
 	// Generate and gossip a new version certificate
-	newVersionCertificate, err := sb.generateVersionCertificate(version)
+	newVersionCertificate, err := istanbul.NewVersionCertificate(version, sb.Sign)
```

Comment:
Its a bit weird to me that this `New*` function takes a signers. It is also different than other messages IIUC. Is there a reason this not done is a separate Sign function, as `istanbul.Message` implements https://github.com/celo-org/celo-blockchain/blob/948566e0d22fbb047be512d53c5e3918b0a19826/consensus/istanbul/types.go#L491?

Reply:
The reason for providing the sign function at construction time is that, we never want an unsigned message. If we didn't sign it at construction time then it opens up the possibility that we could gossip or store an unsigned message and also the possibility that we could waste resources signing a message that has already been signed.

The separate Sign function on istanbul.Message also exhibits these problems, so it would be good to make signing part of construction there as well, and from a quick inspection of the code looks like it would not be so difficult to accomplish.

Reply:
It doesn't seems so crazy to be that construction and signing are done in two steps. Transactions are a great example of why that is useful, and we have proposed more abstract methods of signing, such as utilizing a threshold scheme for each validator. So I personally favor the two step approach, but only weakly.